libmscore/undo.cpp

```diff
@@ -1899,6 +1899,42 @@ void ChangeNoteEvents::flip(EditData*)
       */
       }
 
+//---------------------------------------------------------
+//   ChangeNoteEventList::flip
+//---------------------------------------------------------
+
+void ChangeNoteEventList::flip(EditData*)
+      {
+      note->score()->setPlaylistDirty();
+      // Get copy of current list.
+      NoteEventList nel = note->playEvents();
+      // Replace current copy with new list.
+      note->setPlayEvents(newEvents);
+      // Save copy of replaced list.
+      newEvents = nel;
+      // Get a copy of the current playEventType.
+      PlayEventType petval = note->chord()->playEventType();
+      // Replace current setting with new setting.
+      note->chord()->setPlayEventType(newPetype);
+      // Save copy of old setting.
+      newPetype = petval;
+      }
+
+//---------------------------------------------------------
+//   ChangeNoteEventList::flip
+//---------------------------------------------------------
+
+void ChangeChordPlayEventType::flip(EditData*)
```

Comment:
It looks like this operation is not always truly undoable. If you set play events type to `PlayEventType::Auto` then all the play events will be just overwritten and the user-defined events will be lost. If this type gets set to `InvalidUser` then events seem to change in an undoable way but `ChangeEventList` command seems to [revert](https://github.com/musescore/MuseScore/blob/693d08eb597aeef485575ebec48d4c54c5887c03/libmscore/undo.cpp#L2083-L2087) the play events type back to `PlayEventType::User`.

https://github.com/musescore/MuseScore/blob/693d08eb597aeef485575ebec48d4c54c5887c03/libmscore/rendermidi.cpp#L2385-L2392

If we are going to make use of it in plugins than this should probably be fixed somehow, perhaps saving the old note events could be a solution. Maybe exposing only a subset of such changes (e.g. only setting the `playEventType` to `InvalidUser` to force the events regeneration or resetting it to `Auto` while saving the user-defined event lists) could be exposed for now if that turns out to be easier to implement.

Reply:
This whole InvalidUser thing is confusing. I wasn't even going to allow it through the API. The only place I could see it was set was in the file loader code. Can you shed some light on this?

-Dale

Reply:
Discussion needed.

Reply:
> The only place I could see it was set was in the file loader code.

Could you please point me to that place? Currently I don't see any code which sets the play events type to `InvalidUser` at all.

The intention of introducing `InvalidUser` is understandable though: there should be a way to revert from the user-defined play events to automatic ones in a way that would allow such an operation. But in this case it looks like the solution could be just to rework `ChangeChordPlayEventType` command to store the user-defined events when switching play events type from `User` to `Auto`, then we could get rid of `InvalidUser` type at all.

Reply:
I was mistaken. It is only used in rendermidi.cpp but not set.

Reply:
>The intention of introducing InvalidUser is understandable though: there should be a way to revert from the user-defined play events to automatic ones in a way that would allow such an operation. 

But `User` appears to do the same thing but without the additional undo stuff.

Reply:
Observations of a PlayEventType:

* Only set to `PlayEventType::Auto` in Chord constructor.
* `PlayEventType::Auto` triggers all playEvents generation on the Note level.
* Writes to XML output only occur on when `chord.playEventType` is non-`Auto`.
* `playEventType` is only granular to the Chord level so its value affects ALL note's playEvents.
* `PlayEventType::User` is set if the "Events" XML tag is encountered on loads.
* `Note::SetOn/OffTimeOffset()` sets chord's `playEventType` to `PlayEventType::User` without undo support. They appear to only be called by `static void readNote(Note* note, XmlReader& e)`.
* Set to `PlayEventType::User` in `ChangeEventList`, `ChangeNoteEventList`, `ChangeNoteEvent` constructors when creating Undo objects. These object's capture the current `chord.playEventType` while setting the new value to `User` as supplied by the constructor.
* `PlayEventType` values is in `Ms::PluginAPI::Chord::setPlayEventType()` with undo support.
* The enum `PlayEventType::InvalidUser` is only checked during Midi rendering and is never set in the code.

That's everything I could find. With that in hand...

If we ignore the `PlayEventType::InvalidUser` which is never set in the code, it appears to me that the undo support properly unwinds and rewinds the `Chord::playEventType` when performed atomically. Note that I did have to fix ChangeNoteEvent because the `flip()` override originally only set `PlayEventType::User` but didn't save and restore its previous value. The code had a TODO note on it so I figured it was an unfinished programming thought.

Reply:
I mean that if you have a chord with `PlayEventType::User` and want to switch it to `PlayEventType::Auto` in a way that would allow undoing this operation then you will have also to save the play events that this chord have had before switching the play events type. Otherwise those old (potentially user-edited) events will be completely lost as they will be silently overwritten on the next play events creation operation. If this approach gets implemented then I see no reason to have that `InvalidUser` type anymore.

Reply:
Ahhhh.... Now I get it! 

That must be what the undo class `ChangeNoteEvents` was intended for. 

So what we need is to have the ChangeChordPlayEventType add a vector of playlists one per note, to stash the state of those items. Yes?

There may be a helper in the chord for constructing that as well.

-Dale

Reply:
> So what we need is to have the `ChangeChordPlayEventType` add a vector of playlists, one per note, to stash the state of those items. Yes?

Yes, that's exactly what I mean.

Reply:
I'm on it then!
-Dale

Reply:
New code submitted.